src/metabase/api/channel.clj

```diff
@@ -23,21 +24,27 @@
   "Test if a channel can be connected, throw an exception if it fails."
   [type details]
   (try
-    (let [result (channel/can-connect? (models.channel/keywordize-type type) details)]
+    (let [result (channel/can-connect? type details)]
       (when-not (true? result)
         (throw (ex-info "Unable to connect channel" (merge {:status-code 400} result)))))
     (catch Exception e
       (throw (ex-info "Unable to connect channel" (merge {:status-code 400} (ex-data e)))))))
 
+(def ^:private ChannelType
+  ;; TODO can we add a decoder that turn a string to a keyword?
+  (mu/with-api-error-message
```

Comment:
@escherize, is there a way to make this schema decode to `keyword` automatically?

Reply:
I think you want something like this (warning used an LLM)

```clojure
(def ChannelType
  [:and
   [:or
    [:string {:coerce keyword}]
    [:keyword]]
   [:fn (fn [x] (= (namespace x) "channel"))]])
```

Reply:
Seems it works with [small tweak](https://malli.io/?value=%22channel%2Fwhat%22&schema=%5B%3Aand%0A%20%5B%3Aor%0A%20%20%5B%3Astring%20%7B%3Acoerce%20keyword%7D%5D%0A%20%20%5B%3Akeyword%5D%5D%0A%20%5B%3Afn%20(fn%20%5Bx%5D%20(%3D%20(namespace%20(keyword%20x))%20%22channel%22))%5D%5D)

Reply:
You could do it that way, or you could just put the coercion data onto the fn form itself. I'm not exactly sure what shapes you want to allow, but the way I'd mess with it in the repl is:

``` clojure

#_:clj-kondo/ignore ;;nocommit
(require '[malli.core :as mc] '[malli.error :as me] '[malli.util :as mut] '[metabase.util.malli :as mu]
         '[metabase.util.malli.describe :as umd] '[malli.provider :as mp] '[malli.generator :as mg]
         '[malli.transform :as mtx] '[metabase.api.common.internal :refer [defendpoint-transformer]])

(def my-schema [:fn {:decode/string (fn [x] (keyword x))}
                #(= "channel" (namespace (keyword %)))])

(mc/decode my-schema "channel/what" defendpoint-transformer)
;; => :channel/what


(map (fn [v] [v (mc/validate [:fn {:decode/string (fn [x] (keyword x))}
                             #(= "channel" (namespace (keyword %)))]
                            v)])
     ["channel"
      "channel/what"
      "channel/what/else"
      :channel
      :channel/who])
;; => (["channel" false]
;;     ["channel/what" true]
;;     ["channel/what/else" true]
;;     [:channel false]
;;     [:channel/who true])

```

Reply:
Isn's this lovely, thank you gentlemen! 6df536ff49